Document .Net core version binding

BillWagner opened the pull request:
Fixed #2880 

The first commit adds the design document without changes. The second and third commit represent edits from spec to docs.

BillWagner committed:
initial commit

Original design doc.
Updated TOC
Add paragraph in overview topic

BillWagner committed:
first draft complete

BillWagner committed:
proofread and grammar check

mairaw reviewed (changes requested):
Thanks for working on this @BillWagner! Got halfway through this so far. Left some comments.

mairaw commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,170 @@
+
+## Publish a self-contained application
+
+You can publish an application as a **self-contained distribution**. This approach includes .NET Core. Self-contained distributions do not have a dependency on runtime environments. Runtime binding occurs at publishing time, not run-time.
```
link to https://docs.microsoft.com/en-us/dotnet/core/deploying/#self-contained-deployments-scd for the definition of self-contained and the other anchor framework-dependent apps. Also, do we need to link from that topic back here for learning about version binding at publishing time?

KathleenDollard replied:
in second sentence: "includes NET Core runtime."

Drop the third sentence. It is incorrect. The fourth is the point. If you like, move the third to fourth and restate to be very clear you mean it does not depend on a framework that is installed on the machine.

BillWagner commented on docs/core/versions/binding.md:
````diff
@@ -0,0 +1,170 @@
+<TargetFramework>netcoreapp2.0</TargetFramework>
+```
+
+You set multiple target frameworks if you need to build different code for different environments. Setting multiple target frameworks is more common for libraries, but can be done with applications as well. You specify a `TargetFrameworks` property (plural of `TargetFramework`). The target frameworks will be semicolon-delimited as shown in the following example:
````
Me either. I tried ".net implementations" but that doesn't feel right either.

BillWagner commented on docs/core/versions/binding.md:
````diff
@@ -0,0 +1,170 @@
+<TargetFrameworks>netcoreapp2.0;net47</TargetFrameworks>
+```
+
+A given SDK will support a fixed set of frameworks, typically capped to the target framework of the runtime(s) it includes. For example, the .NET Core 2.0 SDK includes the .NET Core 2.0 runtime, which is an implementation of the `netcoreapp2.0` target framework. The .NET Core 2.0 SDK will support `netcoreapp1.0`, `netcoreapp1.1`, and `netcoreapp2.0` but not `netcoreapp2.1` (or higher). You'll nee ...
````
Future tense: I caught some. Checking again.

KathleenDollard replied:
Did we earlier draw the relationship between a target framework and a runtime?

At what point prior to this did we explain that the SDK contains a runtime? Let's drop them in gently by ending the last sentence (and no plural here that I can think of)

...capped to the target framework of the runtime it ships with. For example...

BillWagner commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,170 @@
+
+You take advantage of the latest SDK features and improvements while targeting earlier .NET Core versions. You can target multiple versions of .NET Core on different projects, using the same SDK tools for all projects.
+
+When needed, you configure `dotnet` to use a different version of the SDK. You specify that version in a [global.json file](../tools/global-json.md). The "use latest" policy means you only use `global.json` to specify a .NET Core version earlier than the latest installed version.
```
I think yes. @richlander do you agree?

KathleenDollard commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,170 @@
+.NET Standard target frameworks are also capped in the same way. The .NET Core 2.0 SDK will be capped to `netstandard2.0`.
+
+A given SDK defines a fixed minimum runtime patch version for each target framework that it supports. The minimum runtime patch version is maintained at the `.0` version (for example, `2.0.0`) for the lifetime of a major/minor runtime version family.
+
```
I'm not sure this makes sense. You say it defines a fixed minimum patch, but the patch could never be less than zero. Not sure the point of this paragraph. If it's to set up the next para, I think that could be more clear.

BillWagner replied:
dropped it. It only seemed to add confusion

KathleenDollard commented on docs/core/versions/binding.md:
````diff
@@ -0,0 +1,170 @@
+<RuntimeFrameworkVersion>2.0.4</RuntimeFrameworkVersion>
+```
+
+When you target multiple target frameworks, you must specify the minimum runtime patch version to the specific .NET Core target framework, as shown in the following example:
````
I didn't know this. 

@livarcocc can you confirm?

KathleenDollard commented on docs/core/versions/binding.md:
````diff
@@ -0,0 +1,170 @@
+<RuntimeFrameworkVersion Condition="'$(TargetFramework)' == 'netcoreapp2.0'">2.0.4</RuntimeFrameworkVersion>
+```
+
+A given SDK supports a fixed set of target frameworks. Typically, the target frameworks are the SDK framework version and a set of earlier versions.
````
Two things...

There seems an implication here that the frameworks are only .NET Core frameworks. 

@livarcocc don't we allow targeting all previous .NET Core frameworks?

BillWagner replied:
I'll assume @KathleenDollard is correct that all previous .NET Core faremworks are allowed. @livarcocc please correct if necessary.

KathleenDollard commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,170 @@
+You run an application from source with `dotnet run`. `dotnet run` both builds and runs an application.
+
+The built application runs on machines that satisfy the minimum required runtime patch version, stored in the `*.runtimeconfig.json` file. The highest patch version is selected. For example, the `2.0.4` runtime is selected in the case that the minimum runtime version is `2.0.0` while `2.0.0`, `2.0.1` and `2.0.4` are all installed on the machine. Higher minor or major versions (for example, `2.1.1 ...
+
```
I find it this paragraph that is confusing. You say here that higher versions won't be considered, but in the next paragraph you say they are. 

@livarcocc Do we have major roll forward in any SDK?

KathleenDollard commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,170 @@
+
+You run an application from source with `dotnet run`. `dotnet run` both builds and runs an application.
+
+The built application runs on machines that satisfy the minimum required runtime patch version, stored in the `*.runtimeconfig.json` file. The highest patch version is selected. For example, the `2.0.4` runtime is selected in the case that the minimum runtime version is `2.0.0` while `2.0.0`, `2.0.1` and `2.0.4` are all installed on the machine. Higher minor or major versions (for example, `2.1.1 ...
```
Throughout this part with all the version numbers, I'd like to ensure that it is clear that we are talking about runtime version. It is clear in this paragraph, so I wanted to illustrate what I meant. Later in the examples, it is not.

KathleenDollard commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,170 @@
+- It's possible that 2.0.5 and 2.2.2 might behave differently, particularly for scenarios like serializing binary data.
+
+NuGet dependencies are treated differently. An application might be built for .NET Core 2.0 but run on .NET Core 2.1 due to this algorithm. There may be later versions of an application's NuGet dependencies that target .NET Core 2.1. These later versions aren't considered or installed. All NuGet dependencies are resolved as part of the publish operation and reside as assemblies (.dlls) in a flat  ...
+
```
I'd like to drop the first sentences, and start with All NuGet...

KathleenDollard commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,170 @@
+NuGet dependencies are treated differently. An application might be built for .NET Core 2.0 but run on .NET Core 2.1 due to this algorithm. There may be later versions of an application's NuGet dependencies that target .NET Core 2.1. These later versions aren't considered or installed. All NuGet dependencies are resolved as part of the publish operation and reside as assemblies (.dlls) in a flat  ...
+
+Occasionally, ASP.NET packages are deployed via a web host rather than with a published application. In those cases, the web host should correctly configure their environments based on published guidance such that all supported ASP.NET applications run correctly.
+
```
Not sure this para adds anything

BillWagner commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,170 @@
+
+The built application runs on machines that satisfy the minimum required runtime patch version, stored in the `*.runtimeconfig.json` file. The highest patch version is selected. For example, the `2.0.4` runtime is selected in the case that the minimum runtime version is `2.0.0` while `2.0.0`, `2.0.1` and `2.0.4` are all installed on the machine. Higher minor or major versions (for example, `2.1.1 ...
+
+It's an error if the minimum version specified for an application is not satisfied. However, you may run an application on a machine that is missing the runtime required by the application. If a higher minor version is installed, the minor version is used instead of failing. Higher major versions will not be used to load applications. This behavior is referred to as "minor version roll-forward."
```
This paragraph was redundant with the previous paragraph. Combined where applicable, deleted the rest.

BillWagner committed:
respond to feedback

KathleenDollard commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,170 @@
+You can publish an application as a **self-contained distribution**. This approach includes .NET Core. Self-contained distributions do not have a dependency on runtime environments. Runtime binding occurs at publishing time, not run-time.
+
+The publishing process will select the latest patch version of the given runtime family. For example, `dotnet publish` will select .NET Core 2.0.4 if it is the latest patch version in the .NET Core 2.0 runtime family, independent of the existence of earlier patch versions or minimum runtime patch version values. The publishing process differs from `dotnet build` runtime selection for two reasons:
+
```
drop "independent of...values." Part is redundant, part incorrect

KathleenDollard commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,170 @@
+
+The publishing process will select the latest patch version of the given runtime family. For example, `dotnet publish` will select .NET Core 2.0.4 if it is the latest patch version in the .NET Core 2.0 runtime family, independent of the existence of earlier patch versions or minimum runtime patch version values. The publishing process differs from `dotnet build` runtime selection for two reasons:
+
+- The application doesn't specify a hosting environment.
```
I don't know what is intended here. The published package is specifies it's platform, which is how I would have read this first bullet

BillWagner replied:
dropped that list.

KathleenDollard commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,170 @@
+The publishing process will select the latest patch version of the given runtime family. For example, `dotnet publish` will select .NET Core 2.0.4 if it is the latest patch version in the .NET Core 2.0 runtime family, independent of the existence of earlier patch versions or minimum runtime patch version values. The publishing process differs from `dotnet build` runtime selection for two reasons:
+
+- The application doesn't specify a hosting environment.
+- The developer generates a final configuration for the application (that can't be practically modified). The best runtime choice is the latest installed runtime patch version. It's the same version used for development and it has the latest (per the machine) security and reliability fixes.
```
Let's simplify this

KathleenDollard commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,170 @@
+- The application doesn't specify a hosting environment.
+- The developer generates a final configuration for the application (that can't be practically modified). The best runtime choice is the latest installed runtime patch version. It's the same version used for development and it has the latest (per the machine) security and reliability fixes.
+
+The `runtimeframeworkversion` element will override the default version policy if that element is present in the project file.
```
Please move before intent where the rules are listed.

KathleenDollard commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,170 @@
+
+The `runtimeframeworkversion` element will override the default version policy if that element is present in the project file.
+
+It is an error if the minimum version specified for an application is not satisfied. `dotnet publish` binds to latest runtime patch version (within a given major.minor version family). `dotnet publish` does not support the roll-forward semantics of `dotnet run`.
```
Second sentence has already been stated. Third is problematic. It does support patch roll-forward, as already discussed. The rules are different, which if it's going to be stated should probably be in the first paragraph of the self-contained section. 

@dsplaisted can you confirm that at publish there is not runtime minor roll-forward?

KathleenDollard commented on docs/core/versions/binding.md:
````diff
@@ -0,0 +1,170 @@
+You publish with the latest runtime patch version for the given target framework specifying `--latest`. This argument has the same default runtime binding behavior for publishing self-contained applications. See the following example:
+
+``` console
+dotnet publish -c release -o app --latest
````
what is the value of the other switches here. Either drop or write out as the reader may not remember what c and o stand for

BillWagner replied:
dropping this section

KathleenDollard commented on docs/core/versions/binding.md:
````diff
@@ -0,0 +1,170 @@
+You publish with a specific runtime patch version for the given target framework specifying `--version=[version]`. It is an error if this version is missing. It's an error if the given version doesn't support the target framework (as a lower bound). You may specify higher versions (even major versions), provided they exist on the machine. You can see this usage in the following example:
+
+``` console
+dotnet publish -c release -o app --version=3.1.0
````
Please tell me that syntax is incorrect. Space not equal is the normal syntax.

KathleenDollard replied:
We should probably use a version that we have even contemplated

BillWagner replied:
@KathleenDollard I took that paragraph from the design spec: dotnet/design2#3  It looks like this entire section (Publish an application with a specific .NET Core version) isn't implemented at all. I'm dropping this entire paragraph. Let me know if that's incorrect.

KathleenDollard commented on docs/core/versions/index.md:
```diff
@@ -27,6 +27,16 @@ With .NET Core 2.0, downloads show a single version number in their file name. T
 
 The use of a single version number makes it easier for users to know what version of the SDK to install on their dev machines, and what the corresponding version of the shared framework should be when time comes to provision a production environment. When downloading an SDK or runtime, the version number you see is going to be the same.
```
just a heads up. I think this is the article I'm rewriting. 

"version manager" feels like quite a stretch for our experience. Since I want to build that, I'd rather change hte first bullet here

dsplaisted replied:
Yeah I think this whole document needs to be reviewed / rewritten.  It sounds like Kathleen is doing so.

To be clear, there is **not** a single version number for both the runtime and the SDK :)

KathleenDollard reviewed (commented):
Overall I think this article takes a simple topic and makes it more complicated than it needs to be. 

Also: 

- inconsistent use of the words runtime and framework. I think we should make sure we have some clear definitions on that
- use of the word runtime to mean both .NET Core runtime and the running time of an application
- a lack of clarity brought on by the use of .NET Core (which is simply a brand, there is no such actual thing) to refer to the SDK or runtime)

BillWagner commented:
Added the WIP label. Once @livarcocc and @dsplaisted comment, I'll update based on Kathleen's comments and get it ready for final review.

KathleenDollard commented:
[We made a change to global.json behavior in 2.1](https://github.com/dotnet/core-setup/pull/4118)

This pull request explains the behavior for global.json overall. 

The change was what happened if an SDK version (let's say 2.0.3) was selected and that version was not on the machine, but a lower version (let's say 2.0.2) was the highest version number on the machine. Prior to 2.1 2.0.2 would run. After 2.1 this is an error. We perceived this as a bug fix. 

@johnbeisner can you confirm that this fix went in to 2.1.0 (and that I'm right it's a runtime deployment). 

This is host behavior which we install with the runtime, so you will get this behavior with .NET Core Runtime 2.1.0 or higher, regardless of the highest SDK you have. This last sentence might be too much detail for the topic. 

---
Update the SDK muxer behavior; modify the SDK muxer tests.

**Prior SDK muxer behavior:**
- If no _global.json_ file is found, or the _global.json_ doesn't contain a specified SDK version, the latest installed SDK version is used. Latest SDK version can be either release or pre-release - the highest version number wins.
- If *global.json* does contain a specified SDK version:
  - If the specified SDK version is found on the machine, that exact version is used.
  - If the specified SDK version cannot be found on the machine, the latest installed SDK **patch-version** of the specified SDK version is used. Latest installed SDK **patch-version** can be either release or pre-release - the highest version number wins. 
  - If the specified SDK version and an appropriate SDK **patch-version** cannot be found, an error is thrown.

**Newer SDK muxer behavior:**
- If no _global.json_ file is found, or the _global.json_ doesn't contain a specified SDK version, the latest installed SDK version is used. Latest SDK version can be either release or pre-release - the highest version number wins.
- If *global.json* does contain a specified SDK version:
  - If the specified SDK version is found on the machine, that exact version is used.
  - If the specified SDK version cannot be found on the machine, the latest installed SDK **patch-version** of the specified SDK version which is semantically greater than the specified SDK version is used. Selected SDK **patch-version** can be either release or pre-release.
  - If the specified SDK version and an appropriate SDK **patch-version** cannot be found, an error is thrown.

BillWagner committed:
fix merge

BillWagner committed:
fix a bad merge

BillWagner committed:
address detailed comments.

BillWagner committed:
one more round of detailed changes.

BillWagner committed:
rework focus

Reworked the focus based on feedback. This should be simpler, more scenario focused, and more clear for users.

BillWagner commented:
Removed the 'wip' label.  This is ready for another review.  I took @KathleenDollard 's comments about simplicity and focus to update the headers, explain the behavior from a developer / user perspective. I think this reads much more clear.

@dsplaisted @livarcocc I **think** I distilled the roll forward behavior correctly. I would appreciate a review for accuracy.

dsplaisted commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,117 @@
+
+[!INCLUDE [topic-appliesto-net-core-2plus](../../../includes/topic-appliesto-net-core-2plus.md)]
+
+This article explains the policies used by the .NET Core tools, SDK, and runtime for selecting versions. These policies provide a balance between always running applications using the specified versions and enabling ease of upgrading both developer and end user machines. These policies perform the following:
```
> balance between always running applications

Remove "always" here.

dsplaisted commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,117 @@
+
+## The SDK uses the latest installed version
+
+SDK commands include `dotnet new`, `dotnet build` or `dotnet run`. The `dotnet` CLI must choose an SDK version for any command. The .NET Core CLI uses the latest SDK installed on the machine. You'll use the .NET Core 99.9 SDK when it's installed, even if the project you are working with targets the .NET Core Runtime 2.0. Note that this is true for preview versions as well as released versions. Yo ...
```
Consider using an example with real version numbers here:

> You'll use the .NET Core SDK v2.1.301 if it's the highest version installed, even if the project you are working with targets the .NET Core Runtime 2.0.

dsplaisted commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,117 @@
+
+.NET Standard target frameworks are also capped in the same way. The .NET Core 2.0 SDK is capped to `netstandard2.0`.
+
+You can override the minimum runtime patch version (to higher or lower versions) in the project file, as shown in the following example:
```
It feels like an abrupt jump to start talking about `RuntimeFrameworkVersion` here.  Perhaps this belongs later in the document (after the roll forward and possibly self-contained discussion?)

dsplaisted replied:
It's also important to note that `RuntimeFrameworkVersion` does not specify a minimum patch version, at least for self-contained apps (which is probably the main scenario for using it).  For self-contained apps, the exact version specified will be used, even if there is a later patch version available.

BillWagner replied:
I moved this section to the end of the article.

dsplaisted commented on docs/core/versions/binding.md:
````diff
@@ -0,0 +1,117 @@
+<RuntimeFrameworkVersion Condition="'$(TargetFramework)' == 'netcoreapp2.0'">2.0.4</RuntimeFrameworkVersion>
+```
+
+## dotnet run rolls forward
````
I think this section should probably not be specific to `dotnet run`.  It should be for running a framework-dependent app, which can be via `dotnet run` or via `dotnet app.dll`.  In the .NET Core SDK 2.1.400, [we will support](https://github.com/dotnet/cli/issues/6237#issuecomment-398181320) framework-dependent deployments which include an executable which can be launched natively.  So for Windows you will be able to run them by launching `app.exe`, and on other OS's, you would just run `app`.

BillWagner replied:
I changed the header to use the term "Framework-dependent apps" Is that what we want to standardize on?

dsplaisted commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,117 @@
+
+You run an application from source with [`dotnet run`](../tools/dotnet-run.md). `dotnet run` both builds and runs an application.
+
+The .NET CLI chooses the latest patch version installed on the machine. For example, if you specified `netcoreapp2.0` in your project file, and `2.0.4` is the latest .NET runtime installed, the `2.0.4` runtime is used.
```
I wouldn't say it's the .NET CLI that chooses this.  Usually when we say CLI we mean the .NET Core SDK, but this logic is part of the host, and applies even if you have only the runtime but not the SDK installed.

BillWagner replied:
I also added a definition of **host**. I tried to future-proof it a bit, but I think it will need a small update when [this](https://github.com/dotnet/cli/issues/6237#issuecomment-398181320) arrives.

dsplaisted commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,117 @@
+
+The .NET CLI chooses the latest patch version installed on the machine. For example, if you specified `netcoreapp2.0` in your project file, and `2.0.4` is the latest .NET runtime installed, the `2.0.4` runtime is used.
+
+If no acceptable `2.0.*` version is found, a new `2.*` version will be used. For example, if you specified `netcoreapp2.0` and only `2.1.0` is installed, the application will run using the `2.1.0` runtime. This This behavior is referred to as "minor version roll-forward." Lower versions also won't be considered. When no acceptable runtime is installed, the application will not run.
```
There is a duplicate "This" here

dsplaisted commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,117 @@
+A few usage examples demonstrate the behavior:
+
+- 2.0.4 is required. 2.0.5 is the highest patch version available. 2.0.5 is used.
+- 2.0.4 is required. No 2.0.* versions are installed. 1.1.1 is the highest runtime installed. An error message is printed.
```
Consider "displayed" instead of "printed"

dsplaisted commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,117 @@
+- 2.0.4 is required. 2.0.5 is the highest patch version available. 2.0.5 is used.
+- 2.0.4 is required. No 2.0.* versions are installed. 1.1.1 is the highest runtime installed. An error message is printed.
+- 2.0.4 is required. No 2.0.* versions are installed. 2.2.2 is the highest 2.x runtime version installed. 2.2.2 is used.
+- 2.0.4 is required. No 2.x versions are installed. 3.0.0 is installed. An error message is printed.
```
Consider "displayed" also here.

dsplaisted commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,117 @@
+- 2.0.5 is later installed. 2.0.5 will be used for subsequent application launches, not 2.2.2.
+- It's possible that 2.0.5 and 2.2.2 might behave differently, particularly for scenarios like serializing binary data.
+
+All NuGet dependencies are resolved as part of the publish operation and reside as assemblies (.dlls) in a flat directory structure. These assemblies should run on .NET Core 2.1 due to its compatibility promise for existing applications and binaries.
```
Does this sentence serve a purpose?  I would delete it.  It's not entirely accurate for packages which have assemblies that are also part of .NET Core, or part of the ASP.NET Core Shared Framework.

BillWagner replied:
removed.

dsplaisted commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,117 @@
+
+All NuGet dependencies are resolved as part of the publish operation and reside as assemblies (.dlls) in a flat directory structure. These assemblies should run on .NET Core 2.1 due to its compatibility promise for existing applications and binaries.
+
+Developers should remember the following rules:
```
I'm not sure it's helpful to restate the logic here.  I would consider deleting form here through the end of the section.

BillWagner replied:
removed as well.

dsplaisted commented on docs/core/versions/binding.md:
```diff
@@ -0,0 +1,117 @@
+
+The publishing process selects the latest patch version of the given runtime family. For example, `dotnet publish` will select .NET Core 2.0.4 if it is the latest patch version in the .NET Core 2.0 runtime family. The target framework (including the latest installed security patches) are packaged with the application.
+
+It is an error if the minimum version specified for an application is not satisfied. `dotnet publish` binds to latest runtime patch version (within a given major.minor version family). `dotnet publish` does not support the roll-forward semantics of `dotnet run`.
```
I'd suggest linking from here to https://docs.microsoft.com/en-us/dotnet/core/deploying/runtime-patch-selection, which includes more details on roll-forward for self-contained publish.

dsplaisted reviewed (commented):
This looks pretty good overall.  However, I wonder whether "binding" is the right term to use to describe the behavior.  It feels like a pretty technical term that we're trying to use in docs for an audience that won't generally know what "binding" means in this context.

Could we call it "version selection" instead of "binding"?

BillWagner committed:
updated per product team feedback.

BillWagner commented:
Thanks for the comments @dsplaisted That cleared up my remaining questions.

@KathleenDollard I did not update versions/index.md as you are rewriting that now.  Will you be addressing [this comment](https://github.com/dotnet/docs/pull/6060#discussion_r198665106)?

BillWagner committed:
fix link

BillWagner commented:
@mairaw I believe all the changes have been addressed. Can you give this a final review?

dsplaisted commented on docs/core/versions/index.md:
```diff
@@ -27,6 +27,15 @@ With .NET Core 2.0, downloads show a single version number in their file name. T
 
 The use of a single version number makes it easier for users to know what version of the SDK to install on their dev machines, and what the corresponding version of the shared framework should be when time comes to provision a production environment. When downloading an SDK or runtime, the version number you see is going to be the same.
 
+### Version binding
```
It looks like renamed the document from "Binding" to "Version Selection", but the term hasn't been updated everywhere.  Is this intentional?

KathleenDollard replied:
As I was looking at @dsplaisted 's comment, I noticed 

"The use of a single version..."

would this read better as 

"The use of the same version number" 
or
"Using the same version number"

BillWagner replied:
@dsplaisted Got it. I had thought you meant to update "binding" for SEO, not because we weren't using the term.

@KathleenDollard Let's handle that with the next PR, based on #6081  The SDK and Runtime versions aren't aligned, so that section will change a lot.

dsplaisted commented on docs/core/deploying/index.md:
```diff
@@ -36,7 +36,7 @@ There are also a few disadvantages:
 ## Self-contained deployments (SCD)
 
-For a self-contained deployment, you deploy your app and any required third-party dependencies along with the version of .NET Core that you used to build the app. Creating an SCD doesn't include the [native dependencies of .NET Core](https://github.com/dotnet/core/blob/master/Documentation/prereqs.md) on various platforms, so these must be present before the app runs.
+For a self-contained deployment, you deploy your app and any required third-party dependencies along with the version of .NET Core that you used to build the app. Creating an SCD doesn't include the [native dependencies of .NET Core](https://github.com/dotnet/core/blob/master/Documentation/prereqs.md) on various platforms, so these must be present before the app runs. For more information on vers ...
```
Change link to selection.md, and consider changing text from "binding" to "version selection" (twice).

rpetrusha replied:
nit: comma between "runtime" and "see"

dsplaisted commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+- Easy and efficient deployment of .NET Core, including security and reliability updates.
+- Usage of the latest tools and commands independent of target runtime.
+
+There are four activities during development and deployment where version binding occurs:
```
binding -> selection

dsplaisted commented on docs/core/versions/index.md:
```diff
@@ -27,6 +27,15 @@ With .NET Core 2.0, downloads show a single version number in their file name. T
 
+### Version binding
+
+.NET Core applies a set of policies that determine which versions of the .NET Core runtime and SDK are used in various scenarios. These scenarios and policies are fully explored in the article on [binding](selection.md).
```
binding -> version selection

dsplaisted commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+
+## Self-contained deployments include the selected runtime
+
+You can publish an application as a [**self-contained distribution**](../deploying/index.md#self-contained-deployments-scd). This approach includes .NET Core. Self-contained distributions do not have a dependency on runtime environments. Runtime binding occurs at publishing time, not run-time.
```
binding -> version selection

dsplaisted commented on docs/toc.md:
```diff
@@ -120,7 +120,7 @@
 ### [Live unit testing .NET Core projects with Visual Studio](/visualstudio/test/live-unit-testing-start)
 ## [Versioning](core/versions/index.md)
-<!--### [🔧 Servicing](core/versions/servicing.md)-->
+### [.NET Core version binding](core/versions/selection.md)
```
binding -> selection

BillWagner committed:
binding -> version

Per Product team feedback.

rpetrusha commented on docs/core/versions/index.md:
```diff
@@ -27,6 +27,15 @@ With .NET Core 2.0, downloads show a single version number in their file name. T
+
+.NET Core applies a set of policies that determine which versions of the .NET Core runtime and SDK are used in various scenarios. These scenarios and policies are fully explored in the article on [version selection](selection.md).
+
+One can think of these policies performing the following roles:
```
One --> You
policies performing --> policies as performing

rpetrusha commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+This article explains the policies used by the .NET Core tools, SDK, and runtime for selecting versions. These policies provide a balance between running applications using the specified versions and enabling ease of upgrading both developer and end user machines. These policies perform the following:
+
+- Easy and efficient deployment of .NET Core, including security and reliability updates.
+- Usage of the latest tools and commands independent of target runtime.
```
nit: Usage --> Use

rpetrusha commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+- Easy and efficient deployment of .NET Core, including security and reliability updates.
+- Usage of the latest tools and commands independent of target runtime.
+
+There are four activities during development and deployment where version selection occurs:
```
Perhaps just simply, "Version selection occurs:"

rpetrusha commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+- When you [run an SDK command](#the-sdk-uses-the-latest-installed-version).
+- When you [build an assembly](target-framework-monikers-define-build-time-apis).
+- When you [run a .NET Core application](framework-dependent-apps-roll-forward).
+- When you publish a [self-contained application](self-contained-deployments-include-the-selected-runtime).
```
This is confusing, because the descriptions don't correspond to the headers, and these categories aren't mutually exclusive. For example, running an SDK command (item 1) includes building an assembly (item 2).

rpetrusha commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+
+## The SDK uses the latest installed version
+
+SDK commands include `dotnet new`, `dotnet build` or `dotnet run`. The `dotnet` CLI must choose an SDK version for any command. The .NET Core CLI uses the latest SDK installed on the machine. You'll use the .NET Core SDK v2.1.301 when it's installed, even if the project you are working with targets the .NET Core Runtime 2.0. Note that this is true for preview versions as well as released versions ...
```
Best to qualify  "the latest SDK" with "by default"

rpetrusha commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+
+SDK commands include `dotnet new`, `dotnet build` or `dotnet run`. The `dotnet` CLI must choose an SDK version for any command. The .NET Core CLI uses the latest SDK installed on the machine. You'll use the .NET Core SDK v2.1.301 when it's installed, even if the project you are working with targets the .NET Core Runtime 2.0. Note that this is true for preview versions as well as released versions ...
+
+On rare occasions, you may need to use a different version of the SDK. You specify that version in a [global.json file](../tools/global-json.md). The "use latest" policy means you only use `global.json` to specify a .NET Core version earlier than the latest installed version.
```
Although it's clarified in the second sentence, I'd change "different version" to "earlier version"

rpetrusha commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+
+On rare occasions, you may need to use a different version of the SDK. You specify that version in a [global.json file](../tools/global-json.md). The "use latest" policy means you only use `global.json` to specify a .NET Core version earlier than the latest installed version.
+
+`global.json` can be placed anywhere in the file hierarchy. The CLI searches upward from the project directory for the first `global.json` it finds. You control which projects a given `global.json` applies to by its place in the file system. The .NET CLI searches for a `global.json` file iteratively navigating the path upward from the current working directory. The first `global.json` file found  ...
```
This is partly redundant -- there are two sentences about the CLI searching upward. 
if you keep the fourth sentence, "iteratively navigating" --> iteratively by navigating"

rpetrusha commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+
+The process for selecting an SDK version is:
+
+- `dotnet` searches for a `global.json` file iteratively reverse-navigating the path upward from the current working directory.
```
These could be numbered rather than list items.

rpetrusha commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+- 2.0.4 is required. 2.0.5 is the highest patch version available. 2.0.5 is used.
+- 2.0.4 is required. No 2.0.* versions are installed. 1.1.1 is the highest runtime installed. An error message is displayed.
+- 2.0.4 is required. No 2.0.* versions are installed. 2.2.2 is the highest 2.x runtime version installed. 2.2.2 is used.
+- 2.0.4 is required. No 2.x versions are installed. 3.0.0 is installed. An error message is displayed.
```
perhaps note in parens that there is no 3.0.0?

rpetrusha commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+
+## Self-contained deployments include the selected runtime
+
+You can publish an application as a [**self-contained distribution**](../deploying/index.md#self-contained-deployments-scd). This approach includes .NET Core. Self-contained distributions do not have a dependency on runtime environments. Runtime version selection occurs at publishing time, not run-time.
```
deployment, not distribution
"This approach includes .NET Core" is unclear. "...bundles the .NET Core libraries with your application"?

mairaw replied:
nit: do not have -> don't have

run-time -> at run time

rpetrusha commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+
+The publishing process selects the latest patch version of the given runtime family. For example, `dotnet publish` will select .NET Core 2.0.4 if it is the latest patch version in the .NET Core 2.0 runtime family. The target framework (including the latest installed security patches) are packaged with the application.
+
+It is an error if the minimum version specified for an application is not satisfied. `dotnet publish` binds to latest runtime patch version (within a given major.minor version family). `dotnet publish` does not support the roll-forward semantics of `dotnet run`. For more details see the article on [runtime patch selection](../deploying/runtime-patch-selection.md) in deploying .NET Core applicatio ...
```
latest runtime --> the latest runtime

mairaw replied:
nits; It is -> It's
is not -> isn't
does not -> doesn't

For more details ... -> use the boilerplate sentence "For more information about x, see y."

rpetrusha reviewed (changes requested):
This looks really good, @BillWagner. I've left a number of comments and suggestions.

mairaw commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+
+A given SDK supports a fixed set of frameworks, capped to the target framework of the runtime it ships with. For example, the .NET Core 2.0 SDK includes the .NET Core 2.0 runtime, which is an implementation of the `netcoreapp2.0` target framework. The .NET Core 2.0 SDK will support `netcoreapp1.0`, `netcoreapp1.1`, and `netcoreapp2.0` but not `netcoreapp2.1` (or higher). You install the .NET Core ...
+
+.NET Standard target frameworks are also capped in the same way. The .NET Core 2.0 SDK is capped to `netstandard2.0`.
```
not sure if I follow here. 2.1 SDK is capped to netstandard2.0 as well. Should we say that versions don't need to match between standard and core?

BillWagner replied:
I re-wrote it to add clarity.

mairaw commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+
+## Framework-dependent apps roll forward
+
+You run an application from source with [`dotnet run`](../tools/dotnet-run.md). `dotnet run` both builds and runs an application. The `dotnet` executable is the **host** for the application.
```
Make it clear this is for dev environments?

mairaw commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+
+The host chooses the latest patch version installed on the machine. For example, if you specified `netcoreapp2.0` in your project file, and `2.0.4` is the latest .NET runtime installed, the `2.0.4` runtime is used.
+
+If no acceptable `2.0.*` version is found, a new `2.*` version is used. For example, if you specified `netcoreapp2.0` and only `2.1.0` is installed, the application runs using the `2.1.0` runtime. This behavior is referred to as "minor version roll-forward." Lower versions also won't be considered. When no acceptable runtime is installed, the application will not run.
```
nit: will not run -> won't run

mairaw commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+
+A few usage examples demonstrate the behavior:
+
+- 2.0.4 is required. 2.0.5 is the highest patch version available. 2.0.5 is used.
```
available or installed?

mairaw commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+A few usage examples demonstrate the behavior:
+
+- 2.0.4 is required. 2.0.5 is the highest patch version available. 2.0.5 is used.
+- 2.0.4 is required. No 2.0.* versions are installed. 1.1.1 is the highest runtime installed. An error message is displayed.
```
do we also need an example with 2.0.0 installed?

mairaw commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+Minor version roll-forward has one side-effect that may affect end users. Consider the following scenario:
+
+- 2.0.4 is required. No 2.0.* versions are installed. 2.2.2 is installed. 2.2.2 is used.
+- 2.0.5 is later installed. 2.0.5 will be used for subsequent application launches, not 2.2.2.
```
not 2.2.2 anymore?

BillWagner replied:
Yes, that's correct: it uses the patch instead of an updated minor version.

mairaw replied:
That was a suggestion 😊

mairaw commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+
+The `RuntimeFrameworkVersion` element  overrides the default version policy. For self-contained deployments, the `RuntimeFrameworkVersion` specifies the *exact* runtime framework version. For framework dependent applications, the `RuntimeFrameworkVersion` specifies the *minimum* patch level of the framework.
+
+When you target multiple frameworks, you must specify the minimum runtime patch version to the specific target framework, as shown in the following example for .NET Core:
```
what happens if you don't?

BillWagner replied:
Apparently, it just works. (Spec is no longer correct).  Deleting this paragraph.

mairaw reviewed (approved):
Looks good @BillWagner. Leaving some comments for you to address before merging. Don't forget to check the build report.

BillWagner commented on docs/core/versions/selection.md:
```diff
@@ -0,0 +1,110 @@
+The process for selecting an SDK version is:
+
+- `dotnet` searches for a `global.json` file iteratively reverse-navigating the path upward from the current working directory.
+- `dotnet` uses the SDK specified in the first `global.json` found.
```
I added that link, but I may remove if it causes a build error until #4704 is merged.

pkulikov commented on docs/core/versions/selection.md:
````diff
@@ -0,0 +1,110 @@
+<TargetFramework>netcoreapp2.0</TargetFramework>
+```
+
+You may build your project against multiple TFMs. Setting multiple target frameworks is more common for libraries but can be done with applications as well. You specify a `TargetFrameworks` property (plural of `TargetFramework`). The target frameworks will be semicolon-delimited as shown in the following example:
````
is -> are?

BillWagner committed:
update per feedback.

This can get added back later, once #4704 is approved.

BillWagner commented:
@rpetrusha Thanks for all the feedback (and @mairaw )  

I've addressed all the comments, and this should be ready to go.

BillWagner commented:
closing & reopening for wip check

rpetrusha reviewed (approved):
This looks really good, @BillWagner. You can merge when you're ready.

rpetrusha commented:
closing and reopening so that PR (hopefully) will be mergeable

pkulikov commented:
The last commit message contains "WIP". That seems to be blocking.

rpetrusha commented:
I think, @BillWagner, that the PR is blocked because of your last commit message.
